### PR TITLE
added bit array library

### PR DESCRIPTION
**Pull request by agoryelov**

### CODE REVIEW

**Review by KarelChanivecky — CHANGES_REQUESTED** [review]
- ONE_BIT can be made const
- ONE_BIT is uint8_t but bit getter and setter use bool. Revise consistency to the data type of your choice.
- check return of malloc and calloc. For consistency it would be best to just stick to calloc
- use CHAR_BIT (limits.h) instead of 8
- create header comment that explicitly state endianness and includes a visual example 
- rename struct bit_array to struct bit_array_s to match naming convention( simpler version identifier is reserved for the pointer as it will most likely be used the most)
- add "typedef struct bit_array_s * bit_array" to decrease verbosity in code that uses this data structure
- consider defining constants for each error and returning the errors where appropriate

**Comment by agoryelov** [comment]
Updated based on feedback

File: libs/bit_array.c [hunk 1]
```diff
@@ -4,50 +4,78 @@
...
+    byte += bit_array_check_bit(this, index++) * 16;
+    byte += bit_array_check_bit(this, index++) * 8;
+    byte += bit_array_check_bit(this, index++) * 4;
+    byte += bit_array_check_bit(this, index++) * 2;
+    byte += bit_array_check_bit(this, index++) * 1;
+
+    return byte;
+}
+
+void bit_array_add_bit(bit_array this, bool bit_value) {
```
Comment (KarelChanivecky) [inline libs/bit_array.c]: please return error code

File: libs/bit_array.c [hunk 2]
```diff
@@ -58,7 +86,7 @@ void bit_array_add_bit(struct bit_array * this, bool bit_value) {
     }
 }
 
-void bit_array_set_bit(struct bit_array * this, size_t index, bool bit_value) {
+void bit_array_set_bit(bit_array this, size_t index, bool bit_value) {
```
Comment (KarelChanivecky) [inline libs/bit_array.c]: Please return error code

File: libs/bit_array.c [hunk 3]
```diff
@@ -4,50 +4,78 @@
...
-void bit_array_destroy(struct bit_array ** p_this) {
+void bit_array_destroy(bit_array * p_this) {
+    free((*p_this)->array);
     free(*p_this);
     *p_this = NULL;
 }
 
-void bit_array_init(struct bit_array * this, size_t initial_bit_capacity) {
-    size_t byte_capacity = (initial_bit_capacity + 7) / 8;
+void bit_array_init(bit_array this, size_t initial_bit_capacity) {
```
Comment (KarelChanivecky) [inline libs/bit_array.c]: Check result of calloc, return success or failure

File: libs/bit_array.c [hunk 4]
```diff
@@ -4,50 +4,78 @@
...
+    byte += bit_array_check_bit(this, index++) * 2;
+    byte += bit_array_check_bit(this, index++) * 1;
+
+    return byte;
+}
+
+void bit_array_add_bit(bit_array this, bool bit_value) {
     if (this->bit_length >= this->bit_capacity) {
-        fprintf(stderr, "bit_array_add_bit: BitArray capacity reached.\n");
+        fprintf(stderr, "bit_array_add_bit: byte_array capacity reached.\n");
```
Comment (KarelChanivecky) [inline libs/bit_array.c]: Please no side effects

**Review by KarelChanivecky — CHANGES_REQUESTED** [review]

**Review by KarelChanivecky — APPROVED** [review]

**Pull request merged**